cmd/habitat/community/state/transitions.go

```diff
@@ -378,3 +381,119 @@ func (t *StopProcessInstanceTransition) Validate(oldState *community.CommunitySt
 
 	return fmt.Errorf("process instance with matching process and node IDs not found")
 }
+
+type AddImplementationTransition struct {
+	InterfaceHash     string
+	DatastoreID       string
+	ContentIdentifier string
```

Comment:
wait why is there a content identifier for the transition? i was thinking we just have the hash + datastore ID and then content identifier gets sent for reads.

Reply:
Good point, this made me think about the architecture here. The content identifier is how a datastore internally addresses the implementation data. Ideally, the client querying Dex never knows about this CID or its format. So I think the core question is "why should we map the interface to the CID in the state machine?"  

The alternative is to have this mapping in the datastore itself. The driver would be responsible for querying this mapping and then getting the impl using the resulting CID from query #1. One huge advantage of doing it this way is that we wouldn't have to propagate changes to the state machine whenever the CID changes (this is not a case we've spent a lot of time thinking about I think). There is a tradeoff with complexity in the driver and roundtrips per Dex query that we need to consider.

I think this is a really interesting tradeoff, we probably should discuss this next time we meet.